cloudfiles/interfaces.py

```diff
@@ -150,92 +169,139 @@ def put_file(
       raise ValueError("Compression type {} not supported.".format(compress))
 
     if (
-      content 
-      and type(content) is str 
-      and content_type 
+      content
+      and type(content) is str
+      and content_type
       and re.search('json|te?xt', content_type)
     ):
 
       content = content.encode('utf-8')
 
-    if hasattr(content, "read") and hasattr(content, "seek"):
-      with open(path, 'wb') as f:
-        shutil.copyfileobj(content, f)
-      return
+    def do_put_file():
+      if hasattr(content, "read") and hasattr(content, "seek"):
+        with open(path, 'wb') as f:
+          shutil.copyfileobj(content, f)
+        return
 
-    try:
-      with open(path, 'wb') as f:
-        f.write(content)
-    except IOError as err:
-      mkdir(os.path.dirname(path))
-      with open(path, 'wb') as f:
-        f.write(content)
+      try:
+        with open(path, 'wb') as f:
+          f.write(content)
+      except IOError as err:
+        mkdir(os.path.dirname(path))
+        with open(path, 'wb') as f:
+          f.write(content)
+
+    if self._lock_dir:
+      lock_path = os.path.join(self._lock_dir,
+                               os.path.abspath(path).replace(os.sep, "=+"))
+      rw_lock = fasteners.InterProcessReaderWriterLock(lock_path)
+      with rw_lock.write_lock():
+        do_put_file()
+    else:
+      do_put_file()
 
   def head(self, file_path):
     path = self.get_path_to_file(file_path)
 
     path, encoding = self.get_encoded_file_path(path)
 
-    try:
-      statinfo = os.stat(path)
-    except FileNotFoundError:
-      return None
+    def do_head():
+      try:
+        statinfo = os.stat(path)
+      except FileNotFoundError:
+        return None
 
-    return {
-      "Cache-Control": None,
-      "Content-Length": statinfo.st_size,
-      "Content-Type": None,
-      "ETag": None,
-      "Last-Modified": datetime.utcfromtimestamp(statinfo.st_mtime),
-      "Content-Md5": None,
-      "Content-Encoding": encoding,
-      "Content-Disposition": None,
-      "Content-Language": None,
-      "Storage-Class": None,
-      "Request-Charged": None,
-      "Parts-Count": None,
-    }
+      return {
+        "Cache-Control": None,
+        "Content-Length": statinfo.st_size,
+        "Content-Type": None,
+        "ETag": None,
+        "Last-Modified": datetime.utcfromtimestamp(statinfo.st_mtime),
+        "Content-Md5": None,
+        "Content-Encoding": encoding,
+        "Content-Disposition": None,
+        "Content-Language": None,
+        "Storage-Class": None,
+        "Request-Charged": None,
+        "Parts-Count": None,
+      }
+
+    if self._lock_dir:
+      lock_path = os.path.join(self._lock_dir,
+                               os.path.abspath(path).replace(os.sep, "=+"))
```

Comment:
Since this is used twice, might be better to make this a function. I suspect if we run into a problem it'll get fixed in one place and not in the other.